File: chorum_murohc/api/progression.py
"""Progression endpoints: the caller's own level, and its one write.

Two routes under `/api/v1/`, both same-origin session authenticated,
child-only, both scoped to the one household the caller is a live child of,
exactly as `api/balances.py` scopes its own two routes:

- `GET progression/` returns the caller's own level state: `level`,
  `max_level`, `lifetime_points`, `next_level_threshold`,
  `points_to_next_level`, and `pending_level_up`. Read-only; it writes
  nothing and emits no audit event, because computing a level is not an
  event any more than reading a balance is.
- `POST progression/acknowledge/` is the one write: it tells the server the
  caller has now seen their current level, so it is never shown again. There
  is no body, because there is nothing for a client to supply; the level
  being acknowledged is always the server's own freshly computed answer, per
  the permission matrix's "the client never supplies ... the resulting
  level". A repeat call, or a call with nothing newly reached, changes
  nothing and answers 200 with the same shape.

Authority is never taken from the client. The household comes from
`resolve_active_membership`, never from a body, query string, path or
header, so the answer is always the caller's own figures, and membership is
re-resolved inside the acknowledge write so a membership revoked a moment
ago cannot slip a write through.

Level: issue #61's decided policy. `chorum_murohc.progression.services` owns
the arithmetic and the write; this module owns only the HTTP shape.
"""

from rest_framework.authentication import SessionAuthentication
from rest_framework.exceptions import PermissionDenied
from rest_framework.renderers import JSONRenderer
from rest_framework.response import Response
from rest_framework.views import APIView

from chorum_murohc.api.permissions import PERMISSION_DENIED_DETAIL, IsHouseholdChild
from chorum_murohc.api.session import resolve_active_membership
from chorum_murohc.identity.models import Membership
from chorum_murohc.progression.services import (
    acknowledge_level_up,
    progression_summary_for_user,
)


class _ProgressionAPIView(APIView):
    """Shared authority for both progression routes.

    An unsupported method is refused before authority is considered, so the
    answer to an unlisted method never depends on who is asking.
    """

    renderer_classes = (JSONRenderer,)
    authentication_classes = (SessionAuthentication,)
    permission_classes = (IsHouseholdChild,)

    def initial(self, request, *args, **kwargs):
        if request.method.lower() not in self.http_method_names:
            self.http_method_not_allowed(request)
        super().initial(request, *args, **kwargs)

    def get_permission_household(self, request):
        """The one household the caller may read or act in, or `None`."""
        membership = resolve_active_membership(request.user, request)
        return None if membership is None else membership.household

    def require_child_membership(self):
        """Re-resolve the acting child, for use inside a write transaction."""
        membership = resolve_active_membership(self.request.user, self.request)
        if membership is None or membership.role != Membership.Role.CHILD:
            raise PermissionDenied(PERMISSION_DENIED_DETAIL)
        return membership


class ProgressionView(_ProgressionAPIView):
    """`GET /api/v1/progression/`: the caller's own level state."""

    http_method_names = ('get', 'head', 'options')

    def get(self, request):
        household = self.get_permission_household(request)
        return Response(progression_summary_for_user(household, request.user))


class ProgressionAcknowledgeView(_ProgressionAPIView):
    """`POST /api/v1/progression/acknowledge/`: mark the level as shown."""

    http_method_names = ('post', 'options')

    def post(self, request):
        membership = self.require_child_membership()
        summary, _changed = acknowledge_level_up(membership.household, membership.user)
        return Response(summary)

File: frontend/src/api/progression.ts
/**
 * The progression API client for the merged levelling contract (issue #61).
 *
 * Two routes under `/api/v1/progression/`, child-only both ways:
 *
 * - `fetchProgression` reads the caller's own level state: `level`,
 *   `max_level`, `lifetime_points`, `next_level_threshold`,
 *   `points_to_next_level`, and `pending_level_up`. The level is a function
 *   of lifetime points earned, never a stored counter and never something
 *   the client supplies.
 * - `acknowledgeLevelUp` tells the server the caller has now seen their
 *   current level, so it is shown once and never replayed. There is no body:
 *   the level being acknowledged is always the server's own freshly computed
 *   answer.
 *
 * Same-origin session authentication and CSRF only, matching `api/chores.ts`
 * and `api/submissions.ts` exactly: `ensureCsrfToken` and the CSRF header
 * they export are reused here rather than duplicated. Every response crosses
 * a trust boundary, so a body is validated at runtime before it is trusted.
 */

import { CSRF_HEADER_NAME, ensureCsrfToken } from './session'

export interface Progression {
  level: number
  max_level: number
  lifetime_points: number
  next_level_threshold: number | null
  points_to_next_level: number | null
  pending_level_up: number | null
}

export const PROGRESSION_QUERY_KEY = ['progression'] as const

/**
 * - `forbidden`: the caller is no longer an active child of this household.
 * - `unavailable`: unreachable, timed out, not JSON, or an unexpected status.
 */
export type ProgressionFailureKind = 'forbidden' | 'unavailable'

export class ProgressionRequestError extends Error {
  readonly kind: ProgressionFailureKind

  constructor(kind: ProgressionFailureKind) {
    super(`Progression request failed: ${kind}`)
    this.name = 'ProgressionRequestError'
    this.kind = kind
  }
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value)

const isJsonMediaType = (contentType: string | null) =>
  contentType !== null &&
  /^application\/(?:[a-z0-9!#$&^_.+-]+\+)?json(?:\s*;|$)/i.test(contentType)

const isNonNegativeInteger = (value: unknown): value is number =>
  typeof value === 'number' && Number.isInteger(value) && value >= 0

const isIntegerOrNull = (value: unknown): value is number | null =>
  value === null || (typeof value === 'number' && Number.isInteger(value))

/** The exact six-key shape the merged contract returns. */
export const isProgression = (value: unknown): value is Progression => {
  if (!isRecord(value) || Object.keys(value).length !== 6) {
    return false
  }
  for (const key of [
    'level',
    'max_level',
    'lifetime_points',
    'next_level_threshold',
    'points_to_next_level',
    'pending_level_up',
  ]) {
    if (!Object.hasOwn(value, key)) {
      return false
    }
  }

  return (
    isNonNegativeInteger(value.level) &&
    isNonNegativeInteger(value.max_level) &&
    isNonNegativeInteger(value.lifetime_points) &&
    isIntegerOrNull(value.next_level_threshold) &&
    isIntegerOrNull(value.points_to_next_level) &&
    isIntegerOrNull(value.pending_level_up)
  )
}

const readJsonBody = async (response: Response): Promise<unknown> => {
  if (!isJsonMediaType(response.headers.get('Content-Type'))) {
    return undefined
  }
  try {
    return await response.json()
  } catch {
    return undefined
  }
}

const send = async (path: string, init: RequestInit): Promise<Response> => {
  try {
    return await fetch(path, { credentials: 'same-origin', ...init })
  } catch {
    // The reason is dropped: it can carry a URL or a platform diagnostic, and
    // the interface shows one fixed sentence either way.
    throw new ProgressionRequestError('unavailable')
  }
}

const readProgressionBody = async (response: Response): Promise<Progression> => {
  if (response.redirected) {
    throw new ProgressionRequestError('unavailable')
  }
  if (response.status === 403) {
    throw new ProgressionRequestError('forbidden')
  }
  const body = await readJsonBody(response)
  if (response.status !== 200 || !isProgression(body)) {
    throw new ProgressionRequestError('unavailable')
  }
  return body
}

/** `GET /api/v1/progression/`: the caller's own level state. */
export const fetchProgression = async ({
  signal,
}: { signal?: AbortSignal } = {}): Promise<Progression> =>
  readProgressionBody(
    await send('/api/v1/progression/', {
      method: 'GET',
      headers: { Accept: 'application/json' },
      signal,
    }),
  )

/**
 * `POST /api/v1/progression/acknowledge/`: mark the current level as shown.
 *
 * Idempotent: calling it again with nothing newly reached returns the same
 * shape, `pending_level_up` already `null`, and changes nothing on the
 * server.
 */
export const acknowledgeLevelUp = async ({
  csrfToken,
  signal,
}: {
  csrfToken: string
  signal?: AbortSignal
}): Promise<Progression> =>
  readProgressionBody(
    await send('/api/v1/progression/acknowledge/', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      signal,
    }),
  )

export { ensureCsrfToken }
